### PR TITLE
Add JSON wrapping generic capability

### LINKED ISSUE #1072: Expected `cli.Generic` but actual is `string`
## my urfave/cli version is

v2

## Checklist

* [x] Are you running the latest v2 release? The list of releases is [here](https://github.com/urfave/cli/releases).
* [x] Did you check the manual for your release? The v2 manual is [here](https://github.com/urfave/cli/blob/master/docs/v2/manual.md)
* [x] Did you perform a search about this problem? Here's the [Github guide](https://help.github.com/en/github/managing-your-work-on-github/using-search-to-filter-issues-and-pull-requests) about searching.

## Dependency Management

- Bazel

## Describe the bug

Trying to specify a generic flag to load from yaml.

## To reproduce

Create a generic flag and provide a string value in yaml. 

## Observed behavior

`Expected 'cli.Generic' but actual is 'string'`

## Expected behavior

It works

## Additional context

If the issue relates to a specific open source Github repo, please link that repo here.

If you can reproduce this issue with a public CI system, please link a failing build here.

## Want to fix this yourself?

We'd love to have more contributors on this project! If the fix for this bug is easily explained and very small, free free to create a pull request for it.
## Run `go version` and paste its output here

```
go version go1.13.1 linux/amd64
```

## Run `go env` and paste its output here

```
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/preston/.cache/go-build"
GOENV="/home/preston/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/preston/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build299372313=/tmp/go-build -gno-record-gcc-switches"
```

### PR DESCRIPTION
**Pull request by meatballhat**
## What type of PR is this?

- bug
- feature

## What this PR does / why we need it:

This change adds JSON wrapping generic capability to allow for wrapping of any type with JSON marshal/unmarshal so that it may provide the `Generic` interface. This is also useful when reading values from `altsrc` providers where the type information is likely at a more primitive level than what a given `Generic` type may allow.

## Which issue(s) this PR fixes:

Fixes #1072 

## Special notes for your reviewer:

I'm labeling this both `kind/feature` and `kind/bug` given that I'm addressing a bug fix (#1072) with a feature.

### CODE REVIEW

File: flag_test.go [hunk 1]
```diff
@@ -1310,12 +1312,26 @@ func TestJSONGenericWrapper(t *testing.T) {
...
+				}
+				return &m
+			}(),
+			expectErr: true,
+		},
 	} {
 		t.Run(tc.name, func(ct *testing.T) {
 			fl := JSONWrapGeneric(tc.wrapped)
 
 			if err := fl.Set(fl.String()); err != nil {
```
Comment (dearchap) [inline flag_test.go:1329]: if err == nil and expectErr is true that is a t.Fatal(...). so these kinds of checks have 2 parts

if err != nil && !expectErr {

} else if err == nil && expectErr {

}

File: flag_test.go [hunk 2]
```diff
@@ -1310,12 +1312,26 @@ func TestJSONGenericWrapper(t *testing.T) {
...
-				ct.Errorf("error %v", err)
+				if !tc.expectErr {
+					ct.Errorf("unexpected error %v", err)
+				} else if _, ok := err.(*json.SyntaxError); !ok {
```
Comment (dearchap) [inline flag_test.go:1332]: why do we care if its a json syntaxError or something else ?

**Review by dearchap — COMMENTED** [review]
Please make req changes

**Comment by meatballhat** [comment]
@dearchap hello! Sorry, I should have said something about _why_ I marked this as "Draft" 😅  I'm not expecting feedback yet, because I'm not convinced this is working at all like I'd want it to. If I can get it fixed up I'll bring the PR out of "Draft" and re-request your review 💖

**Comment by meatballhat** [comment]
Nope! I'll revisit in the future if it seems like a reasonable approach.

**Pull request closed without merging**